From: Mike O'Connor <mjo@dojo.mi.org>
Date: Wed, 15 Feb 2006 05:17:24 +0000 (-0500)
Subject: [PATCH] XFS ftruncate() bug could expose stale data (CVE-2006-0554)
X-Git-Url: http://www.kernel.org/git/?p=linux/kernel/git/chrisw/linux-2.6.15.y.git;a=commitdiff;h=8dcd7c19f2624b7150edd60da336da0bb5291bef

[PATCH] XFS ftruncate() bug could expose stale data (CVE-2006-0554)

This is CVE-2006-0554 and SGI bug 942658.  With certain types of
ftruncate() activity on 2.6 kernels, XFS can end up exposing stale
data off disk to a user, putting extents where holes should be.

Signed-off-by: Chris Wright <chrisw@sous-sol.org>
---

--- a/fs/xfs/linux-2.6/xfs_aops.c
+++ b/fs/xfs/linux-2.6/xfs_aops.c
@@ -385,7 +385,7 @@ xfs_probe_unmapped_cluster(
 
 	/* First sum forwards in this page */
 	do {
-		if (buffer_mapped(bh))
+		if (buffer_mapped(bh) || !buffer_uptodate(bh))
 			break;
 		total += bh->b_size;
 	} while ((bh = bh->b_this_page) != head);
